### ISSUE TITLE
Running Reform with dry-validation on top of Rails stack

### ISSUE THREAD

**Issue by blelump** [apotonick/reform#337]
Hi!

I'm trying to abandon ActiveModel and replace it with dry-validation, however still no luck. I've prepared raw Rails app to play with it and extended Gemfile with
```
gem 'dry-validation'
gem 'reform'
```

Then I've created simple form:
```
class UserForm < Reform::Form
  property :email
  validation :default do
    key(:email, &:filled?)
  end
end
```
however it consistently refuses to work, because of `NoMethodError: undefined method 'key' for #<Reform::Form::ActiveModel::Validations::Group:0x00000004cef6b8>`. Yep, I've `class_evaled` the `Form` class as such:
```
require "reform/form/dry"
Reform::Form.class_eval do
  feature Reform::Form::Dry
end
```

What's interesting, without Rails env it plays smoothly, e.g:
```
require 'reform'
require "reform/form/dry"
Reform::Form.class_eval do
  feature Reform::Form::Dry::Validations
end

require_relative 'app/forms/user_form.rb'


form = UserForm.new(OpenStruct.new)
puts form.validate( {:email => ""})
  => false
```

Note that in this case it properly resolves the `Reform::Form::Dry::Validations::Group` class instead of the `ActiveModel` one.

___

### Update

OK, while I believe I've found the problem [here](https://github.com/apotonick/reform/blob/master/lib/reform/rails.rb#L12) ( haven't  noticed it earlier, damnit) I still don't have any explanation why it behaves in a such manner (probably I'm too dumb to understand all the Ruby gotchas, because another `class_eval` should just do the job). Nontheless I believe it should be dropped as the docs [says](https://github.com/apotonick/reform#installation) I should set validator explicitly.

___

### Update 2 

I've digged into it a bit deeper and I'd definitely blame myself , because what for me looks the same, actually isn't. In particular, consider [this](https://github.com/apotonick/reform/blob/master/lib/reform/form/active_model/validations.rb#L13) and [this](https://github.com/apotonick/reform/blob/master/lib/reform/form/dry.rb#L19). These two touch `singleton_class`, however in a way that is a mystery for me. [Here's](http://pastie.org/10719686) a  piece of code that unveils the problem (good catch during an interview ;-).

Anyway, I believe it should be fixed (I'd vote for manually loading what I'm interested instead of 'magic'), so that one is able to use `dry-validation` in the Rails ecosystem.

___

### Update 3

@apotonick , FYI, I've forked the repo and dropped loading of AM by default. Here's the commit https://github.com/blelump/reform/commit/d7edd0137ec2aec4866a6e61826d6dd37778136a , however I am not sure this is enough for a new PR, but on the other hand [README](https://github.com/apotonick/reform#installation) clearly says what have to be done before using AM or dry-validation.

___

### ~~Update 4 :-)~~

~~Despite I've dropped `Reform::Form::ActiveModel::FormBuilderMethods`, the `fields_for` and `accepts_nested_attributes_for`  thing stopped to work, because of missing what's being done [here](https://github.com/blelump/reform/blob/master/lib/reform/form/active_model/form_builder_methods.rb#L37).~~

~~It seems while working with Rails env, it's still essential to include this module, damn it.~~

**Comment by apotonick** [comment]
Use Reform 2.2 from `reform/master`, it will solve all our problems! :stuck_out_tongue_winking_eye:

**Comment by franzejr** [comment]
@apotonick using 2.2 from `reform/mater` and I'm having the same issue.

**Comment by dlupu** [comment]
@apotonick I can confirm what @franzejr has said. We are on  a Rails project that uses reform with Active Model validations. We are in the process of migrating those validations to Dry Validations. During the transition phase we will have some forms that use Dry Validation and others that use Active Model Validations. Here is how we handle this.

**Gemfile**
```ruby
gem 'dry-validation'
gem 'reform', git: 'https://github.com/apotonick/reform.git'
gem 'reform-rails'
```

**config/initializers/reform.rb**
```ruby
require 'reform/form/active_model/validations'
require 'reform/form/dry'

# parent class for all Forms with Active Model Validations
class FormWithActiveModelValidations < Reform::Form 
  include Reform::Form::ActiveModel::Validations
end

# parent class for all Forms with Dry Validations
class FormWithDryValidations < Reform::Form
  feature Reform::Form::Dry
end
```

So now let's take 2 simple forms 

``` ruby
  class DummyQuantityActiveModelValidationsForm < FormWithActiveModelValidations
      property :quantity
      validates :quantity, presence: true
  end

 class DummyQuantityDryValidationsForm < FormWithDryValidations
     property :quantity

    validation :default do
          key(:quantity, &:filled?)
    end
end
```

The class using dry validations will trigger an error `NoMethodError: undefined method 'key' for #<Reform::Form::ActiveModel::Validations::Group>`

The workaround is to add `feature Reform::Form::Dry::Validations` in **each** class that uses dry validation.

```ruby
 class DummyQuantityDryValidationsForm < FormWithDryValidations
     feature Reform::Form::Dry::Validations   # FIXME : hack to make dry-validations work in rails
     property :quantity

    validation :default do
          key(:quantity, &:filled?)
    end
end
```

**Comment by blelump** [comment]
@dlupu , @franzejr 

it's not a hack, I mean you need to put `include Reform::Form::Dry::Validations` into `FormWithDryValidations` class. The docs are just missleading (see also [here](https://github.com/apotonick/reform/blob/master/test/validation/dry_validation_test.rb#L10)).

You guys are familiar with the news that `dry-v` 0.7.0 is going to be released very soon and it will change the DSL? Besides current Reform version is not compatible with the `dry-v` upcoming version. 

I'd suggest to suspend the migration until 0.7 is out and integrated into Reform. Obviously , you can fork and experiment on your own branch.

**Comment by dlupu** [comment]
@blelump ok. thanks for warning us about the next release of dry validations
